internal/js_library/v2/js_library.bzl

```diff
@@ -0,0 +1,172 @@
+  )
+
+def _babel_conversion(ctx, inputs, config, out_dir, mnemonic, description):
+  outputs = _declare_babel_outputs(ctx, out_dir)
+  args = _create_babel_args(ctx, config.path, out_dir)
+  _run_babel(ctx, inputs, outputs, args, mnemonic, description)
+  return outputs
+
+def _es5_conversion(ctx, inputs, config):
+  out_dir = ""
+  return _babel_conversion(ctx, inputs, config, out_dir, "JsCompile", "Babel")
+
+def _collect_sources(ctx, es5_outputs):
+  es5_sources = depset(es5_outputs)
+  es6_sources = depset(ctx.files.srcs)
+  transitive_es5_sources = depset()
+  transitive_es6_sources = depset()
+  for dep in ctx.attr.deps:
+    if hasattr(dep, "typescript"):
+        transitive_es5_sources = depset(transitive = [
+            transitive_es5_sources,
+            dep.typescript.transitive_es5_sources,
+        ])
+        transitive_es6_sources = depset(transitive = [
+            transitive_es6_sources,
+            dep.typescript.transitive_es6_sources,
+        ])
+
+  return struct(
+    es5_sources = es5_sources,
+    transitive_es5_sources = depset(transitive = [transitive_es5_sources, es5_sources]),
+    es6_sources = es6_sources,
+    transitive_es6_sources = depset(transitive = [transitive_es6_sources, es6_sources])
+  )
+
+
+def _js_library(ctx):
+  if ctx.attr.babelrc:
+    config = ctx.file.babelrc
+  else:
+    config = _write_config(ctx)
+
+  inputs = ctx.files.srcs + ctx.files.data + [config]
+
+  es5_outputs = _es5_conversion(ctx, inputs, config)
+
+  js_providers = _collect_sources(ctx, es5_outputs)
+
+  # Return legacy providers as ts_devserver still uses legacy format
+  return struct(
+    typescript = struct(
+      es6_sources = js_providers.es6_sources,
+      transitive_es6_sources = js_providers.transitive_es6_sources,
+      es5_sources = js_providers.es5_sources,
+      transitive_es5_sources = js_providers.transitive_es5_sources,
+    ),
+    legacy_info = struct(
+      files = js_providers.es5_sources,
+      tags = ctx.attr.tags,
+      module_name =  ctx.attr.module_name,
+    ),
+    providers = [
+      DefaultInfo(
+          files = js_providers.es5_sources,
+          runfiles = ctx.runfiles(
+            collect_data = True,
+            collect_default = True,
+          ),
+      ),
+      OutputGroupInfo(
+          es5_sources = js_providers.es5_sources,
+          es6_sources = js_providers.es6_sources,
+      ),
+    ],
+  )
+
+js_library = rule(
+    implementation = _js_library,
+    attrs = {
+        "srcs": attr.label_list(
+            doc = """JavaScript source files from the workspace.
+            These can use ES2015 syntax and ES Modules (import/export)""",
+            allow_files = [".js"]
+        ),
+        "deps": attr.label_list(
+            doc = """Other rules that produce JavaScript outputs, such as `ts_library`.""",
+        ),
+        "data": attr.label_list(
+            doc = """Other files useful for babel such as .browserslistrc""",
```

Comment:
we need some design discussion about whether this rule should leak the fact that babel is used.

Doing so means that users could write things that only work with a given version of babel, or that don't correctly interoperate with other bazel rules.

Generally the philosophy is that the built-in rules provide a much simpler configuration surface and mental model than using the tools directly, and that customization happens through rule composition, not by changing settings to a particular tool. If power users want to have more control, they can write `.bzl` code that extends from these for their purposes (like `ng_rollup_bundle` extends `rollup_bundle`)

Reply:
I think that is a tricky one here, for non-js rules I would totally agree with you, but I think that is what js developers expect. It is one of the first things I hear internally for our rules: how can I pass in my custom eslintrc, how can I pass in my custom nuxt/jest config, etc. So these are power users within the js world but some of them are totally unfamiliar with the internals of bazel, so asking them to learn yet another thing, rather than just passing in a config they are familiar with makes the selling point for bazel much harder.

And to be honest I have been in the same position where I wanted to customize the rollup_bundle and while I theoretically understand the benefits of not exposing custom configs in practice I find it much preferable to have to ensure my config is still working after upgrading as this also means I get to reap any potential new features and don't have to keep my internal rules in sync. And if for some reason `rollup_bundle` would become deprecated I could still go ahead and create an internal rule for it.

It seems we at ecosia are not the only ones who would want this type of customisation in the js world, given there has been some positive response on https://github.com/bazelbuild/rules_nodejs/pull/234 and there are a bunch of issues around customising rollup, there is also https://github.com/bazelbuild/rules_nodejs/pull/365 and it was actually @an-kumar who pushed to allow a custom babel config.

Maybe the naming is part of the issue here? Should we call this `babel_library`?

Also to take this a bit further it seems to me that `rules_typescript` are leaking the fact that they use `tsc` to compile by mentioning and exposing `tsconfig.json`, right? If we where to swap out `tsc` with `babel@7` I would assume all usages around `tsconfig.json` would start breaking.

But does it maybe make sense to have a discussion about that off of github? As it does seem to be a bigger one.

Reply:
I think we should have a living document about the philosophy of the Bazel toolchain, and what things should be the same as a pure JS toolchain and which things are different.

Yes, we have some rules which expose bits of the native config. Protractor is another one (https://angular.github.io/bazel-builds/src/protractor/protractor_web_test.html#protractor_web_test ). In these, we must take some care because parts of the config can be controlled by the user but other parts are related to inputs, outputs, dep graph, and downstream rules and therefore should be managed by bazel. The `--output` flag for Typescript is the obvious one - we output the correct ES syntax level for whatever downstream rule requested the output. Users who want control over that option don't understand how the build graph works. (Not that they should have to understand it, but we don't give them that "foot gun" either).

I agree that if this exposes Babel internals, naming the rule with `babel` in the name is a good convention to indicate that.

Yeah we still need to work through the rollup_bundle customization as well. I think this is a tricky balance of allowing nearly all use cases without making the rules brittle such that they no longer compose easily.

Reply:
Started for now with renaming to `babel_library`. Still happy to have some further discussions.